README.md
# Praetorian

Praetorian is a structured JSON validator.  Taking JSON data payload and a JSON schema, it's primary usage is to ensure the data meets the required specification of said schema.

## Features

* Type validation
* Requirement verification
* Schema prompt generator
* Automatic type conversion
* Multi-lingual support
* Comprehensive documentation

## Philosophy

While there are other JSON schema validators in the wild, often they are complex, un-weildy or require a lot of knowledge upfront.  Praetorian has been designed to be simple to deploy, fast to execute and flexible to support common use cases.

## Installation

```sh
$ npm install praetorian
```

## Schema

### Basic Types

Praetorian supports the following data types "out of the box", `boolean`, `date`, `decimal`, `integer`, `string`.  By defining these in your schema against a JSON key, Praetorian will ensure the value passed in the JSON data is of that type. Types are required.

#### Boolean

```sh
{
  "shield": true
}
```

#### Date

```sh
{
  "shield": "2016-06-30"
}
```

#### Decimal

```sh
{
  "shield": 1.07
}
```

#### Integer

```sh
{
  "shield": 14
}
```

#### String

```sh
{
  "shield": "wooden"
}
```

### Complex Types

Praetorian supports both complex data types `array` and `object`.  Both can be used in a schema recursively.

#### Array

Type `array` needs a JSON key of `items`.  Inside an `array` you can pass more basic or complex types as required.  As with basic types, `array` can have a "required" JSON key.  

_Note: when passing JSON data as an `array` type, the "items" structure can be repeated as many times as you need._

```sh
{
  "weapon": {
    "type": "array",
    "required": true,
    "items": {
      "sword": {
        "type": "string"
      },
      "dagger": {
        "type": "string"
      }
    }
  }
}
```

#### Object

Type `object` needs a JSON key of `properties`.  Inside an `object` you can pass more basic or complex types as required.  As with basic types, `object` can have a `required` JSON key.  

_Note: unlike with `array`, when passing JSON data as type `object`, the `properties` structure can only be included once._

```sh
{
  "weapon": {
    "type": "object",
    "required": true,
    "properties": {
      "sword": {
        "type": "string"
      },
      "dagger": {
        "type": "string"
      }
    }
  }
}
```

### Requireds

To mark a JSON key as a required field simply do the following:

```sh
{
  "shield": {
    "required": true
  }
}
```
To pass validation of a schema with a `required` property, the supplied JSON data must have this field present and correct
_Note: if the field is not required, you can set `"required": false` or simply remove the property entirely._

### Data "cleaning"

During the `praetorian.validate()` call, any data that is not recognised by the schema will be removed from the result.

## Example 1

Using an `array` type.  With the following example notice in the result, the key `location` on the second object in the `senators` array has been stripped out of the "result".  As it doesn't form part of the schema, Praetorian removes this property.

_Schema_
```sh
{
  "senators": {
    "type": "array",
    "items": {
      "name": {
        "type": "string"
      },
      "age": {
        "type": "integer"
      }
    }
  }
}
```

_JSON_
```sh
{
  "senators": [
    {
      "name": "Graccus",
      "age": 68
    },
    {
      "name": "Quintus",
      "age": 56,
      "location": "Rome"
    }
  ]
}
```

_Result_
```sh
{
  "senators": [
    {
      "name": "Graccus",
      "age": 68
    },
    {
      "name": "Quintus",
      "age": 56
    }
  ]
}
```

## Usage

To include Praetorian in your project and get underway, use the following snippet.

```sh
var Praetorian = require( 'praetorian' );
praetorian = new Praetorian( options );

// pass your data and structure in like this
praetorian.validate( json, schema, function( err, result ) {
  if( err ) {
    // in the instance of a schema validation error, requirements will tell you for the passed schema
    // how to fulfill the specification
    praetorian.requirements( schema, function( err, result ) {
      if( err ) {
        // requirements errored
      } else {
        // requirements "result" drop out here
      }
    } );
  } else {
    // clean data "result" drops out here
  }
} );
```

## Options

Options can be passed into the Praetorian constructor to modify the default behaviour like this:

```sh
var Praetorian = require( 'praetorian' );
praetorian = new Praetorian(
  {
    automaticTypeConversion: false,
    language: "es"
  }
);
```

### - Automatic type conversion

By default `automaticTypeConversion` is set to `true`.  When `praetorian.validate()` is called, where "types" are specified in your schema, Praetorian will attempt to convert an obvious types to their native.  To turn this off, set it to `false`.

__How?__
```sh
{
  automaticTypeConversion: false 
}
```

__Schema__
```sh
{
  "hasHorse": {
    "type": "boolean"
  }
}
```

__JSON__
```sh
{
  "hasHorse": "true"
}
```

__Result__
```sh
{
  "hasHorse": true
}
```

### - Language

By default `language` is set to `en` (English).  When `.validate()` or `.requirements()` is called, any messages returned will be in the requested language.  Supported languages can be found in the `lib/internationalisation` folder.

_Note: `description` keys will be returned from the schema as they are passed in._

__How?__
```sh
{
  language: "es"
}
```

__Schema__
```sh
{
  "hasHorse": {
    "type": "boolean",
    "description": "Does this senator have a horse?"
  }
}
```

__JSON__
```sh
{
  "hasHorse": "notABoolean"
}
```

__Result__
```sh
{
  "hasHorse": {
    "example": "debe ser un valor booleano e.g. verdad",
    "description": "Does this senator have a horse?",
    "required": false
  }
}
```

## Testing

To run the test harness change to the praetorian directory:
```sh
cd praetorian
```

and run the tests:
```sh
node test/sanity.js
```

## License

(The [MIT](http://en.wikipedia.org/wiki/MIT_License "MIT") License)

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
## **0.2.1**
Bumped lodash version to 4.5.1
## **0.2.0**
[**#12**](https://github.com/Shortbreaks/praetorian/issues/12) Allow basic arrays for type `array`


test.js
/**
* @name /test.js
* @description Nodejs style test include
* @since Fri Oct 04 16:47:15 BST 2013
* @author Kevin Hodges <kevin.hodges@holidayextras.com>
*/

module.exports = require( './test/sanity' );

test/index.js
// TODO: make the parameters to Praetorian constructor one object with sub properties, not multiple parameters


var crypto = require( 'crypto' );

// using this to build praetorian...
// fire up Praetorian
var Praetorian = require( '../index' );
// as second parameter, pass in custom rules
// I can only use the built in "rules" so we would need to write those ourselves and accept pull requests to the core
// would be nice to be able to add custom rules so its fully customisable...
praetorian = new Praetorian(
	{
		'salt': 'LOL'
	},
	{
		"myCompoundValidator": {
			"rules": {
				"minLength": 3,
				"maxLength": 6
			},
			"rule": "Must between 3 and 6 characters e.g. ABCD"	
		},
		"myInValidator": {
			"rules": {
				"enum": [
					"Y", "N", "T", "A"
				]
			},
			"rule": "Must be one of \"Y\", \"N\", \"T\" or \"A\""
		}
	}
);

var object = {
	"foo": "bar"
};

var json = {
	"fooBar": {
		"value": "he"
	}
};

var schema = {
	"fooBar": {
		"type": "string",
		"validation": "myCompoundValidator"
	}
};

// check shit
praetorian.validate( json, schema, function( err, data ) {

		if( err ) {
			
			console.log( 'check err', err );

			praetorian.requirements( schema, function( err, data ) {

				if( err ) {
					console.log( 'requirements err', err );
				} else {
					console.log( 'requirements success', data );		
				}

			} );
		} else {
			// console.log( 'check success', data );
			console.log( 'data', data );
		}
} );

// {
// 	"myCompoundValidator": {
// 		"rules": {
// 			"minLength": 3,
// 			"maxLength": 6
// 		},
// 		"rule": "Must between 3 and 6 characters e.g. ABCD"	
// 	},
// 	"myInValidator": {
// 		"rules": {
// 			"enum": [
// 				"Y", "N", "T", "A"
// 			]
// 		},
// 		"rule": "Must be one of \"Y\", \"N\", \"T\" or \"A\""
// 	}
// }

// var json = {
// 	"foo": "ABCDEFG",
// 	"bar": "AB",
// 	"baz": "ABCD",
// 	"jon": "A",
// 	"boy": "B",
// 	"roy": 2
// }

// var schema = {
// 	"foo": {
// 		"type": "string",
// 		"validation": "myCompoundValidator",
// 		"required": true,
// 		"description": "this is a foo"
// 	},
// 	"bar": {
// 		"type": "string",
// 		"validation": "myCompoundValidator"
// 	},
// 	"baz": {
// 		"type": "string",
// 		"validation": "myCompoundValidator"
// 	},
// 	"jon": {
// 		"type": "string",
// 		"validation": "myInValidator"
// 	},
// 	"boy": {
// 		"type": "string",
// 		"validation": "myInValidator"
// 	},
// 	"roy": {
// 		"type": "string",
// 		"validation": "myInValidator"
// 	}
// }

// var json = {
// 	"hey": "2014-07-22"
// }

// var schema = {
// 	"hey": {
// 		"type": "date"
// 	}
// }



test/sanity.js
var crypto = require( 'crypto' );

// get praetorian (so we can test it)
var Praetorian = require( '../index' );
praetorian = new Praetorian(
	{
		automaticTypeConversion: true
	}
);

// and some testing stuff...
var vows = require( 'vows' ),
		assert = require( 'assert' );

// hash checking
var falco = require( './json/falco' );
// data verification: ensures you have passed everything required, note we're not 
// doing any type checking or validation here
// do note though, this test should pass with all "requirements" met
var gracchus = require( './json/gracchus' );
// data verification: check that required fields that arent passed in json add an error
// this will throw exactly 8 errors
var gaius = require( './json/gaius' );
// passing type conversion
var lucius = require( './json/lucius' );
// data sanitising: this should strip out all invalid properties
var quintus = require( './json/quintus' );
// failing type conversion
var tiberius = require( './json/tiberius' );
// requirements checker
var tarquinius = require( './json/tarquinius' );

vows.describe( 'praetorian' ).addBatch( {

		'gracchus': {
			topic: function() {
				// fire the call
				praetorian.validate( gracchus.json, gracchus.schema, this.callback );
			},
			'execute gracchus': function( err, result ) {

				// errors thrown by praetorian will be handled and return an err, to pass the value should be null
				assert.isNull( err );

				// check the response isn't null either
				assert.isNotNull( result );

				// must be an object too
				assert.isObject( result );

				// now check that the result json (cleaned data) is EXACTLY what we expect it to be
				assert.deepEqual( result, gracchus.expected );

			}
		},

		'gaius': {
			topic: function() {
				// fire the call
				praetorian.validate( gaius.json, gaius.schema, this.callback );
			},
			'execute gaius': function( err, result ) {

				// no json is passed in, result will be undefined
				assert.isUndefined( result );

				// should be exactly 8 errors, one for each of the required: true root properties
				assert.equal( err.length, 8 );

			}
		},

		'falco': {
			topic: function() {
				// fire the call
				praetorian.validate( falco.json, falco.schema, this.callback );
			},
			'execute falco': function( err, result ) {

				// no json is passed in, result will be undefined
				assert.isUndefined( result );

				// should be exactly 4 errors, all conditions of the has type are invalid
				assert.equal( err.length, 4 );

			}
		},
		
		'lucius': {
			topic: function() {
				// fire the call
				praetorian.validate( lucius.json, lucius.schema, this.callback );
			},
			'execute lucius': function( err, result ) {

				// errors thrown by praetorian will be handled and return an err, to pass the value should be null
				assert.isNull( err );

				// check the response isn't null either
				assert.isNotNull( result );

				// must be an object too
				assert.isObject( result );
				
				// now check that the result json (cleaned data) is EXACTLY what we expect it to be
				assert.deepEqual( result, lucius.expected );

			}
		},

		'quintus': {
			topic: function() {
				// fire the call
				praetorian.validate( quintus.json, quintus.schema, this.callback );
			},
			'execute quintus': function( err, result ) {

				// errors thrown by praetorian will be handled and return an err, to pass the value should be null
				assert.isNull( err );

				// check the response isn't null either
				assert.isNotNull( result );

				// must be an object too
				assert.isObject( result );

				// now check that the result json (cleaned data) is EXACTLY what we expect it to be
				assert.deepEqual( result, quintus.expected );

			}
		},

		'tiberius': {
			topic: function() {
				// fire the call
				praetorian.validate( tiberius.json, tiberius.schema, this.callback );
			},
			'execute tiberius': function( err, result ) {

				// no json is passed in, result will be undefined
				assert.isUndefined( result );

				// should be exactly 3 errors, one for each of the failing boolean conversions
				assert.equal( err.length, 3 );

			}
		},
		
		'tarquinius': {
			topic: function() {
				// fire the call
				praetorian.requirements( tarquinius.schema, this.callback );
			},
			'execute tarquinius': function( err, result ) {

				// errors thrown by praetorian will be handled and return an err, to pass the value should be null
				assert.isNull( err );

				// check the response isn't null either
				assert.isNotNull( result );

				// must be an object too
				assert.isObject( result );
				
				// now check that the result json (cleaned data) is EXACTLY what we expect it to be
				assert.deepEqual( result, tarquinius.expected );

			}
		}

} ).run();

lib/doorman.js
/**
* @name /libs/utils/doorman.js
* @description Validates all inputs against a required structure and returns any errors that it finds
* @since Fri Oct 04 12:20:21 BST 2013
* @author Kevin Hodges <kevin.hodges@holidayextras.com>
*/

(function() {
	// get some stuff we need to support our Doorman
	var _, moment;
	_ = require( 'underscore' );
	moment = require( 'moment' );
	exports.frisk = ( function( data, structure, callback ) {

		try {

			console.log( 'Doorman.validate()' );
			console.log( data );
			console.log( structure );


			// ensure data and structure are both objects
			if( !_.isObject( data ) ) {
				throw 'parameter "data" is not an object';
			} 
			if( !_.isObject( structure ) ) {
				throw 'parameter "structure" is not an object';
			} 

			// loop over all data

			// throw "shit happened";

			callback( null, '200' );


		} catch ( e ) {

			callback( e );

		}
	} );
}).call(this);

lib/praetorian.js
/**
* @name /lib/praetorian.js
* @description Validates all inputs against a required structure and returns any errors that it finds
* @since Fri Oct 04 12:20:21 BST 2013
* @author Kevin Hodges <kevin.hodges@holidayextras.com>
*/

(function() {

	// get some stuff we need to support Praetorian
	var _ = require( 'lodash' ), crypto = require( 'crypto' );

	var praetorian = function( myOptions, myTasks ) {

		var self = this;
		// default options
		self.options = {
			automaticTypeConversion: true,
			language: 'en',
			salt: 'praetorian'
		}

		// if there's any overriding options, blat them over the defaults
		self.options = _.extend( self.options, myOptions );

		// now fetch the language settings based on the requested language
		self.options.messages = require( './internationalisation/' + self.options.language ); // ISO 639‑1

		// tasks are your custom validation routines, make them available to the instance
		self.tasks = myTasks;

	};

	// types we accept
	var primitiveTypes = {
		ARRAY: 'array',
		BOOLEAN: 'boolean',
		DATE: 'date', // ISO 8601, naturally
		DECIMAL: 'decimal',
		HASH: 'hash',
		INTEGER: 'integer',
		OBJECT: 'object',
		STRING: 'string'
	}
	var complexTypes = {
		ITEMS: 'items'
	}

	// create a shorthand then export the Praetorian object for Nodejs
	exports = module.exports = praetorian;

	// publically accessible hash generator
	praetorian.prototype.generateHash = function( data ) {
		return crypto.createHmac( 'sha1', this.options.salt ).update( JSON.stringify( data ) ).digest( 'hex' );
	}

	praetorian.prototype.validate = function( json, schema, callback ) {

		var self = this;

		/*
		*	An array of objects that define all the errors generated by the json against the schema
		*/
		var errors = [];

		var addError = function( element, message ) {
			errors.push( { 'element': element, 'message': message } );
		}

		// little closure to manage rule adherence
		var rules = function( data ) {

			// state for lifetime of the rules call (rules are chainable)
			var check = {
				data: data == null || ( ( isNaN( data ) && data.length == undefined ) && !_.isObject( data ) ) ? '' : data,
				error: null
			}

			return {
				boolean: function() {
					// must be one of these bad boys
					if( _.indexOf( [true, false, 'true', 'false'], check.data ) === -1  ) {
						check.error = self.options.messages.VALUE_NOT_A_BOOLEAN;
					}
					return this;
				},
				date: function() {
					// 1. has the user passed in too many or too few days for the selected month/year combination (factors in leap years)
					// 2. is the month too high or low (>12 or <1)
					// 3. date must be YYYY-MM-DD format
					if( _.isString( check.data ) ) {
						var date = check.data.split( '-' );
						// must have 3 parts
						if( date.length === 3 ) {
							var selectedYear = date[0];
							// parseInt incase we have preceeding zeros (we don't want em)
							var selectedMonth = parseInt( date[1] );
							var selectedDay = parseInt( date[2] );
							//check if data is a vailid date
							if( new Date( selectedYear, selectedMonth, selectedDay ) == 'Invalid Date' ) {
								check.error = self.options.messages.VALUE_FOR_DATE_INVALID_FORMAT;
							}
							// how long is the selected month?
							var daysInMonth = new Date( selectedYear, selectedMonth, 0 ).getDate();

							// have we got too many days?
							if( selectedDay > daysInMonth ) {
								check.error = self.options.messages.VALUE_FOR_MONTH_HAS_TOO_MANY_DAYS;
							} else if( selectedDay < 1 ) {
								// have we got too few days?
								check.error = self.options.messages.VALUE_FOR_MONTH_HAS_TOO_FEW_DAYS;
							}

							// check months is not greater than 12...
							if( selectedMonth > 12 ) {
								check.error = self.options.messages.VALUE_FOR_MONTH_TOO_HIGH;
							} else if( selectedMonth < 1 ) {
								// and not less than 1...
								check.error = self.options.messages.VALUE_FOR_MONTH_TOO_LOW;
							}
						} else {
							// split on hyphen failed to produce 3 parts
							check.error = self.options.messages.VALUE_FOR_DATE_INVALID_FORMAT;
						}
					} else {
						// not a string :/
						check.error = self.options.messages.VALUE_FOR_DATE_INVALID_FORMAT;
					}

					return this;
				},
				decimal: function() {
					// cant assume automaticTypeConversion is on so isNumber will hopefully cover the check here
					if( _.isNaN( parseFloat( check.data ) ) ) {
						check.error = self.options.messages.VALUE_NOT_A_DECIMAL;
					}
					return this;
				},
				hash: function() {
					if( _.isUndefined( check.data.hash ) || ( !_.isString( check.data.hash ) ) ) {
						// invalid or not present "hash"
						check.error = self.options.messages.HASH_INVALID;
					} else if( !_.isObject( check.data.value ) ) {
						// "value" not present or not an object
						check.error = self.options.messages.HASH_VALUE_NOT_A_OBJECT;
					} else {
						// check the hash "checks out"
						if( self.generateHash( check.data.value ) !== check.data.hash ) {
							check.error = self.options.messages.HASH_INVALID;
						}
					}
					return this;
				},
				integer: function() {
					if( _.isNaN( parseInt( check.data ) ) ) {
						check.error = self.options.messages.VALUE_NOT_A_INTEGER;
					}
					return this;
				},
				string: function() {
					if( !_.isString( check.data ) ) {
						check.error = self.options.messages.VALUE_NOT_A_STRING;
					}
					return this;
				},
				maxLength: function( maxLength ) {
					if( check.data.length > maxLength ) {
						check.error = self.options.messages.VALUE_TOO_LONG + ' ' + maxLength + ' ' + self.options.messages.VALUE_CHARACTERS;
					}
					return this;
				},
				minLength: function( minLength ) {
					if( check.data.length < minLength ) {
						check.error = self.options.messages.VALUE_TOO_SHORT + ' ' + minLength + ' ' + self.options.messages.VALUE_CHARACTERS;
					}
					return this;
				},
				enum: function( permittedValuesArray ) {
					if( _.indexOf( permittedValuesArray, check.data ) === -1 ) {
						check.error = self.options.messages.VALUE_NOT_IN_ARRAY;
					}
					return this;
				},
				execute: function( callback ) {
					callback( check.error );
				}
			};
		}

		// if the schema says the value should be boolean, integer or decimal, make sure it is
		var forceType = function( schema, value ) {
			// by default, always pass back what came in so if we can't match it, stuff doesn't go missing
			var result = value;
			switch( schema.type ) {
				case 'boolean':
					// true booleans can stay as they are, if we match string bools, boolean them up bitches
					if ( value === 'true' ) {
						result = true;
					} else if ( value === 'false' ) {
						result = false;
					}
					break;
				case 'integer':
					result = !_.isNaN( parseInt( value ) ) ? parseInt( value ): value;
					break;
				case 'decimal':
					result = !_.isNaN( parseFloat( value ) ) ? parseFloat( value ): value;
					break;
				default:
					break;
			}

			return result;
		}
		/**
		* handleComplexData deals with the complex data type "ITEMS"
		* which is very specific to data type validated on the checksum passed in
		*/
		var handleComplexData = function ( json, schema ) {
			/* sample data structure handled in this function
			json = [
				{
					price: '74.30',
					data: {
						hotel: {
							arrivalDate: '2014-10-28',
							nights: 2,
							ITEMS: { //nested Items(supplement)
								price: '38.20',
								currency: 'GBP',
								data: { supplement: [Object] }
							}
						},
						ticket: [Object]
						},
					checksum: '3b47d34ee00ef9e4bb3aaa746fc487b95b7bb2d6'
				}
			]
			*/
			//loop through the "ITEMS" json
			_.each( json, function( item, key ) {
				//loop through data of each ITEM
				_.each( item.data, function( itemElement, itemKey ) {
					//if there is a schema linked with this item validate
					if( _.has( schema, itemKey ) ) {
						//validate any other schema (other than checksum) specified for this item
						var checkElement = _.cloneDeep( item );
						var elementPrefix = itemKey + ' ';
						parseJson( checkElement, schema[itemKey], elementPrefix );
						requiredJson( checkElement, schema[itemKey], elementPrefix );
					}
					//if data has more IIEMS deal with  them
					if( !_.isUndefined( itemElement.ITEMS ) ) {
						handleComplexData( itemElement.ITEMS, schema );
						//delete the nested ITEMS so we can continue validating the outer ITEMS
						//as the outer checksum calculation would not have included the nested ITEM
						delete itemElement.ITEMS;
						//loop through the items and validate that the hash matches the generated has for the item
					} else {
						if( _.isUndefined( item.checksum ) || item.checksum == '' ) {
							addError( 'ITEMS', self.options.messages.HASH_INVALID )
						} else {
							// copy the object
							var itemCopy = _.cloneDeep( item );
							// get a copy of the checksum
							var oldHash = itemCopy.checksum;
							// now remove it (it doesnt form part of the actual hash )
							delete itemCopy.checksum;
							var newHash = self.generateHash( itemCopy.data );
							// compare the hashes
							if( oldHash !== newHash ) {
								addError( 'ITEMS', self.options.messages.HASH_INVALID );
							}
						}
					}
				} );
			} );
		}

		// check all the fields that are "required" by the schema are present
		// if we hit a complex type (object, array), only recurse if data is passed for a sub-schema
		var requiredJson = function( json, schema, elementPrefix ) {
			// check json also matches the schema requirements
			_.each( schema, function( levelValue, levelKey ) {

				// is this field "required" (explicit true, don't want people passing no 1 or "true" shit)
				if( ( levelValue.required === true ) && ( _.isUndefined( json[levelKey] ) ) ) {
					// BOOO!, #fail
					addError( elementPrefix + levelKey, self.options.messages.FIELD_REQUIRED_BUT_NOT_FOUND );
				} else {
					// "required" is false or undefined, ignore and continue
				}

				// if the json actually has a value...
				if( json[levelKey] ) {

					// We need to ensure that "type" is always passed in.
					if( !levelValue.type ) {
						addError( elementPrefix + levelKey, self.options.messages.TYPE_REQUIRED_BUT_NOT_FOUND );
					} else if( ( !_.isUndefined( levelValue.items ) ) && ( levelValue.type === primitiveTypes.ARRAY ) && _.isArray( json[levelKey] ) ) {
						// schema says type ARRAY and we have "items" or OBJECT and we have "properties" we continue
						_.each( json[levelKey], function( jsonValue, jsonKey ) {
							if( !_.isUndefined( jsonValue ) ) {
								requiredJson( jsonValue, levelValue.items, elementPrefix + levelKey + '[' + jsonKey + '].' );
							} else {
								// in this instance, an element in the array was not passed at all , no harm done
								// i.e. no name[0].initial was passed but name[1].initial was
							}
						} );
					} else if( ( !_.isUndefined( levelValue.properties ) ) && ( levelValue.type === primitiveTypes.OBJECT ) && _.isObject( json[levelKey] ) ) {
						requiredJson( json[levelKey], levelValue.properties, elementPrefix + levelKey + '.' );
					} else if( levelValue.type == complexTypes.ITEMS ) {
						//pass in the json for items and any schema for nested items( levelValue[properties] )
						handleComplexData( _.cloneDeep( json[levelKey] ), levelValue.properties );
					}
				}

			} );

		}

		var parseJson = function( json, schema, elementPrefix ) {
			// for this level, loop over all the json and see if its required by the schema
			_.each( json, function( levelValue, levelKey ) {
				// is it a node in the schema?
				if( !schema[levelKey] ) {
					// it's not, can it
					delete json[levelKey];
				} else {
					// only convert types if the consumer is cool with that
					if( self.options.automaticTypeConversion ) {
						// type conversion for primitive types that might get passed as a string
						json[levelKey] = forceType( schema[levelKey], levelValue );
					}

					// start doing stuff
					if( schema[levelKey].type === primitiveTypes.ARRAY ) {
						/* Array */
						// check its ACTUALLY an array
						if( _.isArray( json[levelKey] ) ) {
							// Does the schema say there should be items?
							if( schema[levelKey].items ) {
								// loop each element in the array and parse those as individual objects against their schema
								_.each( json[levelKey], function( arrayElement, loopCount ) {
									if( !_.isUndefined( arrayElement ) ) {
										parseJson( arrayElement, schema[levelKey].items, levelKey + '[' + loopCount + '].' );
									} else {
										// in this instance, an element in the array was not passed at all , no harm done
										// i.e. name[0].initial was not passed but name[1].initial was
									}
								} );
							} else {
								// Schema says no "items" and as the value is an array, happy with that
							}
						} else {
							// ERROR: this value should be an ARRAY but it's not (object of some other description?)
							addError( elementPrefix + levelKey, self.options.messages.VALUE_TYPE_MUST_BE_ARRAY );
						}
					} else if( schema[levelKey].type === primitiveTypes.OBJECT ) {
						/* Object */
						// must be an object (not null or undefined or anything silly)
						if( _.isObject( json[levelKey] ) && !_.isArray( json[levelKey] ) && !_.isNull( json[levelKey] ) && !_.isUndefined( json[levelKey] ) ) {
							parseJson( json[levelKey], schema[levelKey].properties, levelKey + '.' );
						} else {
							// ERROR: this value should be an OBJECT but its an array
							addError( elementPrefix + levelKey, self.options.messages.VALUE_TYPE_MUST_BE_OBJECT );
						}
					} else if( schema[levelKey].type && schema[levelKey].type !== complexTypes.ITEMS ) {
						// basic "type" validation only non complex primitive types,
						// objects and arrays are dealt with by recursing into them
						var validator = new rules( json[levelKey] );
						validator[schema[levelKey].type]();

						if( schema[levelKey].validation ) {
							if( self.tasks && self.tasks[schema[levelKey].validation] ) {
								// get the task to be executed
								var task = self.tasks[schema[levelKey].validation];
								if( _.isFunction( task ) ) {
									// if its a function, overwrite the task function with the result of that function
									task = task.apply( null, schema[levelKey].arguments || null );
								}
								// loop the rules and queue some ancilliary calls
								_.each( task.rules, function( ruleValue, ruleName ) {
									validator[ruleName]( ruleValue );
								} );
							} else {
								throw new TypeError( self.options.messages.PARAMETER_TASKS_NOT_OBJECT );
							}
						}

						// execute the validation routines and push any errors onto the Praetorian stack
						validator.execute( function( err ) {
							if( err ){
								addError( elementPrefix + levelKey, err );
							}
						} );

					} else {
						// no validation found, whatevs...
					}

				}

			} );

		}

		try {

			// ensure json and schema are both objects
			if( !_.isObject( json ) ) {
				throw new TypeError( self.options.messages.PARAMETER_JSON_NOT_OBJECT );
			}
			if( !_.isObject( schema ) ) {
				throw new TypeError( self.options.messages.PARAMETER_SCHEMA_NOT_OBJECT );
			}
			if( !_.isFunction( callback ) ) {
				throw new TypeError( self.options.messages.PARAMETER_CALLBACK_NOT_FUNCTION );
			}

			parseJson( json, schema, '' );

			requiredJson( json, schema, '' );
			if( errors.length > 0 ) {
				// send the errors back
				callback( errors );
			} else {
				// send the decent json back
				callback( null, json );
			}
		} catch ( err ) {

			// somehow, return the err in the error callback
			if( callback ) {
				callback( err );
			} else {
				return err;
			}

		}

	};

	praetorian.prototype.requirements = function( schema, callback ) {

		try {

			var self = this;

			var addRequirement = function( field ) {
				var requirementRule = self.options.messages.RULE_NOT_GIVEN;
				// returns the field "requirement" details in a consistent schema
				if( ( field.validation ) && self.tasks && ( _.isFunction( self.tasks[field.validation] ) ) ) {
					var taskResult = self.tasks[field.validation].apply( null, field.arguments || null );
					requirementRule = taskResult.rule;
				} else if( ( field.validation ) && self.tasks && ( self.tasks[field.validation].rule ) ) {
					requirementRule = self.tasks[field.validation].rule;
				} else if( ( field.type ) && ( self.options.messages['RULE_MUST_BE_' + field.type.toUpperCase()] ) ) {
					requirementRule = self.options.messages['RULE_MUST_BE_' + field.type.toUpperCase()];
				}
				return {
					'rule': requirementRule,
					'description': ( field['description'] ) ? field['description'] : '',
					'required': ( field['required'] ) ? true : false
				};
			}

			var getRequirements = function( schema, requirements ) {
				var results = {};

				_.each( schema, function( schemaValue, schemaKey ) {
					if( schemaValue.type === primitiveTypes.OBJECT || schemaValue.type === primitiveTypes.ARRAY ) {
						// sub schema depending on the parent type, get the results for it
						var subSchemaResults = {};
						if( schemaValue.properties ) {
							subSchemaResults['properties'] = getRequirements( schemaValue.properties );
						} else {
							subSchemaResults['items'] = getRequirements( schemaValue.items );
						}
						// bind it with the requirements at this level and continue
						results[schemaKey] = _.extend( addRequirement( schemaValue ), subSchemaResults ) ;

					} else {
						// add some details to the results for the schema
						results[schemaKey] = addRequirement( schemaValue );
					}
				} );

				return results;

			}

			if( !_.isObject( schema ) ) {
				throw new TypeError( self.options.messages.PARAMETER_SCHEMA_NOT_OBJECT );
			}
			if( !_.isFunction( callback ) ) {
				throw new TypeError( self.options.messages.PARAMETER_CALLBACK_NOT_FUNCTION );
			}

			// send the results back a-la-node
			callback( null, getRequirements( schema ) );

		} catch ( err ) {
			// somehow, return the err in the error callback
			if( callback ) {
				callback( err );
			} else {
				return err;
			}
		}

	};

} )();